File: leetcode/leetcode_Easy/BinaryTreeInorderTraversal.java
package leetcode_Easy;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class BinaryTreeInorderTraversal {
	public List<Integer> inorderTraversal(TreeNode root){
		List<Integer> lst=new ArrayList<Integer>();
		
		if(root==null) return lst;
		
		Stack<TreeNode> stack=new Stack<TreeNode>();
		
		TreeNode p=root;
		
		while(!stack.empty() || p!=null){
			
			//if it is not null, push to stack
			//and go down the tree to left
			if(p!=null){
				stack.push(p);
				p=p.left;
				
			//if no left child	
			//pop stack,process the node
			//then let p point to the right
			}else{
				TreeNode t=stack.pop();
				lst.add(t.val);
				p=t.right;
			}
		}
		
		return lst;
	}
	
	
}

class Solution{
	
	
	List<Integer> res=new ArrayList<Integer>();
	public List<Integer> inorderTraversal(TreeNode root){
		if(root==null) return res;
		
		inorderTraversal(root.left);
		res.add(root.val);
		inorderTraversal(root.right);
		
		return res;
	}
	
}

File: leetcode/C++/BinaryTreeInorderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int>path;
        stack<TreeNode*>stk;
        while(root!=NULL||!stk.empty())
        {
            while(root!=NULL)
            {
                stk.push(root);
                root=root->left;
            }
            if(!stk.empty())
            {
                root=stk.top();
                path.push_back(root->val);
                stk.pop();
                root=root->right;
            }
        }
        return path;
    }
};